### PR TITLE
Bank Cards now only display messages to the owner if carried

### LINKED ISSUE #40439: Payday messages get displayed to all mobs around
Issue reported from  Round ID: 94311 (/tg/Station Bagil [ENGLISH] [US-EAST] [100% FREE LAG])

Reporting client version: 512

Not sure if this is intended @Iamgoofball

### PR DESCRIPTION
**Pull request by XDTM**
:cl: XDTM
fix: ID Cards now only display payday messages to the owner if they're being carried. Cards on the ground will still display them to nearby mobs.
/:cl:

Fixes #40439

### CODE REVIEW

File: code/modules/economy/account.dm [hunk 1]
```diff
@@ -54,12 +54,16 @@
...
 
-/datum/bank_account/proc/bank_card_talk(fuck)
-	if(!fuck || !bank_cards.len)
+/datum/bank_account/proc/bank_card_talk(message)
+	if(!message || !bank_cards.len)
 		return
 	for(var/obj/A in bank_cards)
-		for(var/mob/M in hearers(1,get_turf(A)))
-			to_chat(M, fuck)
+		if(ismob(A.loc))
```
Comment (Iamgoofball) [inline code/modules/economy/account.dm]: This won't get the loc if it's inside a bag or PDA or box inside a bag, or PDA inside a box inside a bag, I think. Which was stopping me from just doing this before.

Comment (XDTM) [reply code/modules/economy/account.dm]: Oh, i see now. I think someone made a proc for this, i'll see if i can find it.

Comment (Iamgoofball) [reply code/modules/economy/account.dm]: Also, do me a solid and swap it to this:
``		playsound(A, 'sound/machines/twobeep.ogg', 50, 1)
		A.audible_message("[icon2html(A, hearers(A))] *[fuck]*", null, 1)``

Comment (Iamgoofball) [reply code/modules/economy/account.dm]: if not the audible_message part, keep the sound and image, visual/audio feedback is important

Comment (XDTM) [reply code/modules/economy/account.dm]: alright

**Comment by Barhandar** [comment]
stop polishing a turd, at least force the turdmaker to do it

**Comment by XDTM** [comment]
not sure why travis is erroring this now

File: code/__HELPERS/game.dm [hunk 2]
```diff
@@ -494,6 +494,16 @@
...
+	
+//Recursively checks if an item is inside a given type, even through layers of storage. Returns the atom if it finds it.
+/proc/recursive_loc_check(atom/movable/target, type)
+	var/atom/A = target
+	while(TRUE)
+		if(!A.loc)
+			return
+		if(istype(A.loc, type))
+			return A.loc
+		A = A.loc
```
Comment (duncathan) [inline code/__HELPERS/game.dm]: this proc can be replaced more succinctly with 
```
var/atom/A = target
while(A.loc && !istype(A.loc, type))
    A = A.loc
return A.loc
```

File: code/modules/economy/account.dm [hunk 3]
```diff
@@ -54,12 +54,24 @@
...
+		if(card_holder) //If on a mob
+			card_holder.playsound_local(get_turf(card_holder), 'sound/machines/twobeep.ogg', 50, 1)
+			if(card_holder.can_hear())
+				to_chat(card_holder, "[icon2html(A, card_holder)] *[message]*")
+		else if(isturf(A.loc)) //If on the ground
+			for(var/mob/M in hearers(1,get_turf(A)))
+				playsound(get_turf(card_holder), 'sound/machines/twobeep.ogg', 50, 1)
+				A.audible_message("[icon2html(A, hearers(A))] *[message]*", null, 1)
+		else
+			for(var/mob/M in A.loc.contents) //If inside a container with other mobs (e.g. locker)
```
Comment (duncathan) [inline code/modules/economy/account.dm]: the `contents` here is redundant, you can simply loop `for(var/mob/M in A.loc)`

**Review by duncathan — CHANGES_REQUESTED** [review]

**Pull request merged**